### DISCUSSION TITLE
Best-guess track matching based on known number of individuals

### DISCUSSION THREAD

**Discussion by aperkes** [talmolab/sleap#1288] [Ideas]
Apologies if this exists or has already been requested, essentially, I would like an option during tracking to combine all the tracklets into _n_ tracks, (in my case 5) based on the known number of individuals, doing the best it can to keep tracks continuous. 

Some background: 
We have a system where we're tracking fish that have some little refuge areas such that they are out of sight (see picture). There are some other issues with our model being undertrained and some unnecessary obstructions, but at least some of the track-breaks are inevitable, and we're getting thousands of tracks as a result. I know there are only 5 fish in the tank, and I would like to be able to spit out 5 "best-guess" tracks, assembling all the tracklets based on where they start and stop. 

I realize tracklet matching is an entire field unto itself, but putting some of the standard approaches into the GUI would be really helpful, so that you don't have to write code yourself. 

For even more background, I'm trying to make sure our tracking set-up is robust enough that grad students without a lot of coding experience could use it in the future, so that it doesn't all fall apart after I leave. 

![image](https://user-images.githubusercontent.com/6477178/234665458-7a9302c7-a5b4-4800-9641-d8090d76edfa.png)

**Comment by aperkes** [comment]
I realized right after posting that this is a very similar idea to my recently-bumped feature request from ages ago (https://github.com/talmolab/sleap/discussions/797), so I maybe should have just added it to that, but these are actually slightly different ideas, but feel free to combine or close as you see fit.

**Reply by talmo** [reply]
Thanks for linking back to other discussion and for adding the suggestion!

We're currently adding one feature that should partially address this by setting the maximum number of instances during inference regardless of tracking (#1305 and #1306). This will just take the top N instances based on the score.

This still won't be perfect though and can still result in more than N tracks.

We technically have something that should take of this via the [`sleap-track` CLI](https://sleap.ai/guides/cli.html#inference-and-tracking) but I think we're missing something like what you're describing.

I think the thing that would work best would be something like a cap on the track buffer with sticky persistence so we do a "best-effort" type of thing. [Here's a minimal implementation.](https://gist.github.com/talmo/a5415218c194536092cda10f897f8cad)

We'll create an issue to track this feature and update this thread.

**Comment by aperkes** [comment]
@talmo, 
I just realized I never got back to you. I had to track down this thread and use the code you wrote up, since we were again getting some videos that ended up with 1000s of tracks (across 44k frames) which was overloading the RAM when trying to run sleap-convert. Squashing it down to a fixed number of tracks worked great. Thanks for your help!

**Comment by roomrys** [comment]
Linking the open issue: #1308
